### PR TITLE
Template support

### PR DESCRIPTION
**Pull request by sparsh-989**
After this update, Users can select the template of their choice and can render the whole output based on the template of their need. They just need to give the name of template as an input
Screenshot 2024-07-10 at 4 24 22 PM

For NeuronGroup
## Network details

        **Neuron population:**
        - Group {{ neurongrp.name }}, consisting of {{ neurongrp.N }} neurons.

        # Model dynamics
        {% for key, eqn in neurongrp.equations.items() %}
        - $\frac{d}{d t} {{ key }}$ = {{ eqn.expr }}{% if eqn.unit %} [{{ eqn.unit }}]{% endif %}
        {% endfor %}

        {% if neurongrp.user_method %}
        The equations are integrated with the '{{ neurongrp.user_method }}' method.
        {% endif %}

        # Events (if present)
        {% if 'events' in neurongrp %}
        **Events:**

        {% for event, details in neurongrp.events.items() %}
        - If {{ details.threshold.code }}, a {{ event }} event is triggered and {{ details.reset.code }}.
        {% endfor %}
        {% endif %}

        # Constants (if present)
        {% if 'identifiers' in neurongrp %}
        **Constants:**

        {% for identifier, value in neurongrp.identifiers.items() %}
        - {{ identifier }}: {{ value }}
        {% endfor %}
        {% endif %}

        # Initial values (if present)
        {% if 'initializer' in neurongrp and neurongrp['initializer'] %}
        **Initial values:**

        {% for initializer in neurongrp['initializer'] %}
        - {{ initializer.variable }}: {{ initializer.value }}{% if initializer.unit %} [{{ initializer.unit }}]{% endif %}
        {% endfor %}
        {% endif %}

        # Run regularly (if present)
        {% if 'run_regularly' in neurongrp %}
        **Run regularly:**

        {% for run_reg in neurongrp['run_regularly'] %}
        - {{ run_reg }}
        {% endfor %}
        {% endif %}

For Poisson Group

Name {{ poissongrp['name'] }} with population size {{ poissongrp['N'] }} and rate as {{ render_expression(poissongrp['rates']) }}.

        {% if poissongrp.get('identifiers', None) %}
        Constants:
        {{ expand_identifiers(poissongrp['identifiers']) }}
        {% endif %}

        {% if poissongrp.get('run_regularly', None) %}
        Run regularly:
        {% for run_reg in poissongrp['run_regularly'] %}
        {{ expand_runregularly(run_reg) }}
        {% endfor %}
        {% endif %}

### CODE REVIEW

**Comment by mstimberg** [comment]
Following up on today's meeting, here are a few notes about the next steps:

- For testing, use `set_device("markdown", ..., expander=MdExpander(include_monitors=True))`, so that monitors get included in the output as well
- Merge `master` into the branch, and fix the conflicts
- finish up the basic template support first for all the currently supported classes
- make the `template_type` argument optional, and rename the `template1` templates to `default` (maybe also rename `template_type` to `template_name`)
- Add a new option `template_dir` so that the user can provide their own templates. To implement, replace the `PackageLoader` in the `env` variable (should probably be a variable of the class instead of a module variable) by a [`ChoiceLoader`](https://tedboy.github.io/jinja2/generated/generated/jinja2.ChoiceLoader.html), combining a `FileSystemLoader` with the default `PackageLoader`
- replace the `expand_NeuronGroup`, `expand_PoissonGroup`, etc. function by a single function, since they are almost identical now. Something along the lines of:
```python
    def expand_group(self, group, template_name):
        try:
           template = self.env.get_template(template_name)
           md_str = template.render(group=group, expander=self)
        except TemplateNotFound as e:
            raise ValueError(f"Template '{template_name}' not found.")
        return md_str
```
so that e.g. a call to `expand_PoissonGroup(poisson_group)` can be replaced by something like `expand_group(poisson_group, f"PoissonGroup-{template_name}.md")`. Note that this also needs some minor changes in the templates, e.g. `poissongrp` → `group`.

**Review by mstimberg — CHANGES_REQUESTED** [review]

File: brian2tools/mdexport/mdexporter.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ class MdExporter(BaseExporter):
     """
 
     def build(self, direct_call=True, debug=False, expander=None,
-              filename=None, additional_formats=None, template_type=None):
+              filename=None, additional_formats=None, template_type='default', temp_dir_path=None):
```
[the remark this answers is not in the document]
Comment (sparsh-989) [reply brian2tools/mdexport/mdexporter.py]: done

File: brian2tools/templates/EventMonitor-template1.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+| **Monitors**    | **Variables**                                                                                               | **Source**                                           | **Record**                                                                                                 | **Event**                           |
```
[the remark this answers is not in the document]
Comment (sparsh-989) [reply brian2tools/templates/EventMonitor-template1.md]: done

File: brian2tools/templates/SpikeMonitor-table.md [hunk 3]
```diff
@@ -0,0 +1 @@
+{%extends "EventMonitor-template1.md" %}
```
Comment (mstimberg) [inline brian2tools/templates/SpikeMonitor-table.md:1]: This needs to be renamed as well.

**Review by mstimberg — CHANGES_REQUESTED** [review]
One rename is still missing but otherwise it looks good to me!

**Pull request merged**